http-common/src/request.rs

```diff
@@ -186,61 +187,99 @@ where
                 .body(request_body)
                 .expect("cannot fail to create request");
 
-            let err = match tokio::time::timeout(self.timeout, client.request(request)).await {
-                Ok(response) => {
-                    match response {
-                        Ok(response) => break response,
-                        Err(err) => {
-                            if err.is_connect() {
-                                // Network error.
-                                std::io::Error::new(std::io::ErrorKind::NotConnected, err)
-                            } else {
-                                std::io::Error::new(std::io::ErrorKind::Other, err)
-                            }
+            let mut is_throttled = false;
+            let response_future = async {
+                match client.request(request).await {
+                    Ok(response) => {
+                        let (
+                            http::response::Parts {
+                                status: response_status,
+                                headers: response_headers,
+                                ..
+                            },
+                            response_body,
+                        ) = response.into_parts();
+
+                        // Make sure to download body inside the timeout
+                        let response_body = if has_response_body {
+                            let response_body = hyper::body::to_bytes(response_body)
+                                .await
+                                .map_err(|err| Error::new(ErrorKind::Other, err))?;
+
+                            Some(response_body)
+                        } else {
+                            None
+                        };
+
+                        // if response throttled, go into exponential backoff
+                        if response_status == http::StatusCode::TOO_MANY_REQUESTS {
```

Comment:
I think we also discussed using exponential backoff for certain 5xx errors, right?

Reply:
We discussed it, but decided that if hub gives a 500 error we should do the normal retry and fail quickly.